crates/worker/src/partition/snapshot_producer.rs

```diff
@@ -0,0 +1,151 @@
+        response_tx: Option<Sender<anyhow::Result<SnapshotId>>>,
+    ) -> anyhow::Result<()> {
+        let partition_id = self.partition_id;
+
+        let guard = self
+            .latest_snapshot
+            .clone()
+            .try_lock_owned()
+            .context(anyhow!("Snapshot already in progress"))?;
+
+        let partition_store = partition_store.clone();
+        let base_dir = self
+            .config
+            .live_load()
+            .partition_store
+            .snapshots_base_dir()
+            .to_path_buf();
+        let config = self.config.live_load();
+        let cluster_name = config.common.cluster_name().to_string();
+        let node_name = config.common.node_name().to_string();
+
+        restate_core::task_center().spawn_child(
+            TaskKind::PartitionSnapshotProducer,
+            "create-snapshot",
+            Some(self.partition_id),
+            async move {
+                let result = Self::write_snapshot(
+                    cluster_name,
+                    node_name,
+                    partition_store,
+                    partition_id,
+                    base_dir,
+                )
+                .await;
+
+                let mut snapshot = guard;
+                if let Ok(metadata) = &result {
+                    snapshot.replace(metadata.snapshot_id);
+                }
+
+                if let Some(sender) = response_tx {
+                    sender
+                        .send(result.map(|metadata| metadata.snapshot_id))
+                        .ok();
+                }
+                Ok(())
+            },
+        )?;
+
+        Ok(())
+    }
+
+    async fn write_snapshot(
+        cluster_name: String,
+        node_name: String,
+        mut partition_store: PartitionStore,
+        partition_id: PartitionId,
+        snapshot_base_path: PathBuf,
+    ) -> anyhow::Result<PartitionSnapshotMetadata> {
+        let partition_snapshots_path = snapshot_base_path.join(partition_id.to_string());
+        if let Err(e) = std::fs::create_dir_all(&partition_snapshots_path) {
+            warn!(
+                %partition_id,
+                path = ?partition_snapshots_path,
+                error = ?e,
+                "Failed to create partition snapshot directory"
+            );
+            bail!(anyhow!(
+                "Failed to create partition snapshot directory: {:?}",
+                e
+            ));
+        }
+
+        let snapshot_id = SnapshotId::new();
+        let snapshot_path = partition_snapshots_path.join(snapshot_id.to_string());
+        let snapshot = partition_store
+            .create_snapshot(snapshot_path.clone())
+            .await?;
+
+        let snapshot_meta = PartitionSnapshotMetadata {
+            version: SnapshotFormatVersion::V1,
+            cluster_name,
+            partition_id,
+            node_name,
+            created_at: humantime::Timestamp::from(SystemTime::from(MillisSinceEpoch::now())),
```

Comment:
Updated as suggested; I was thinking that some day we should introduce a clock abstraction and it would be easy to find all the places where we get the "current" time but this adds no value now.